nncf/onnx/graph/model_transformer.py

```diff
@@ -411,24 +409,51 @@ def _apply_qdq_node_removing_transformations(
-            for initializer_name in node.input:
-                if initializer_name in initializers:
-                    model.graph.initializer.remove(initializers[initializer_name])
-                if initializer_name in value_infos:
-                    model.graph.value_info.remove(value_infos[initializer_name])
-
-            model.graph.node.remove(node)
+        name_to_node_map = get_name_to_node_map(model)
+        children_node_mapping = get_children_node_mapping(model)
+        # We combine quantize and dequantize nodes into pairs here because it
+        # does not make sense to remove only the quantize node or the dequantize
+        # node. They should be removed together.
+        was_processed = {t.target_point.target_node_name: False for t in transformations}
+        quantize_dequantize_pairs = []
+        for node_name in was_processed:
+            if was_processed[node_name]:
+                continue
+            quantize_node_proto = name_to_node_map[node_name]
+            if quantize_node_proto.op_type != "QuantizeLinear":
+                continue
+            # `quantize_node_proto` has only one child, which is the dequantize node.
+            dequantize_node_proto = next(iter(get_children(quantize_node_proto, children_node_mapping)))
+            assert dequantize_node_proto.op_type == "DequantizeLinear"
+
+            quantize_dequantize_pairs.append((quantize_node_proto, dequantize_node_proto))
+            was_processed[quantize_node_proto.name] = True
+            was_processed[dequantize_node_proto.name] = True
+
+        if not all(was_processed.values()):
+            raise RuntimeError("Invalid transformation commands.")
+
+        initializers = {i.name: i for i in model.graph.initializer}
+        value_infos = {i.name: i for i in model.graph.value_info}
+
+        for quantize_node_proto, dequantize_node_proto in quantize_dequantize_pairs:
+            # Unlink Q-DQ subgraph from graph
+            children = get_children(dequantize_node_proto, children_node_mapping)
+            for child in children:
+                for port_id, input_name in enumerate(child.input):
+                    if input_name == dequantize_node_proto.output[0]:
+                        child.input[port_id] = quantize_node_proto.input[0]
+            # QuantizeLinear and DequantizeLinear nodes have common initializers in ports 1 and 2.
+            for i in [1, 2]:
+                model.graph.initializer.remove(initializers[quantize_node_proto.input[i]])
+                model.graph.value_info.remove(value_infos[quantize_node_proto.input[i]])
+
+            for node_proto in [quantize_node_proto, dequantize_node_proto]:
+                model.graph.value_info.remove(value_infos[node_proto.output[0]])
+
+            model.graph.node.remove(quantize_node_proto)
+            model.graph.node.remove(dequantize_node_proto)
+
+        onnx.checker.check_model(model)
```

Comment:
Why do we have this check?

Reply:
This check failed prior to my changes. So we had an invalid ONNX proto after QDQ removing. I added this check here to be sure that ONNX proto we get her is legal.

Reply:
Why can not we cover this via unit tests?
Could you add a testcase for the previous failed scenario?

Reply:
Could you please explain your concern about this check? It is not clear for me.

> Could you add a testcase for the previous failed scenario?

We have such tests
https://github.com/openvinotoolkit/nncf/blob/c0e51f8833e04697da944745f7488fc4b553f832/tests/onnx/test_model_transformer.py#L197

Reply:
Done